staging/src/k8s.io/legacy-cloud-providers/vsphere/nodemanager.go

```diff
@@ -259,6 +259,16 @@ func (nm *NodeManager) GetNode(nodeName k8stypes.NodeName) (v1.Node, error) {
 	return *node, nil
 }
 
+func (nm *NodeManager) getNodes() map[string]*v1.Node {
+	nm.registeredNodesLock.RLock()
+	defer nm.registeredNodesLock.RUnlock()
+	registeredNodes := make(map[string]*v1.Node, len(nm.registeredNodes))
+	for nodeName, node := range nm.registeredNodes {
+		registeredNodes[nodeName] = node
```

Comment:
Should this be copying the nodes so that we don't mutate them later? Where is registered nodes updated, is this also done under lock?

Reply:
`registeredNodes` returned from here are not updated in the workflow. 

I see in the `(nm *NodeManager) refreshNodes()` we are just iterating over the key `nodeName` and do not update the node object.

we can return copy of nodes instead of a pointer. This will require further refactoring as other methods in the workflow are accepting node pointer as argument.

Reply:
@JoelSpeed Every operation with registeredNodes is now done under lock supervision, and none of them requires special isolation except the one provided by the defined operations like add or delete, which already locking the resource themselves for the further processing. Each resource operating on the node pointer will update the Node object under the pointer, which is the intent of the method - but will do this while locking the needed resource.

Reply:
Just to clarify, what I meant was should we do `registerdNodes[nodeName] = node.DeepCopy()` which would return a copy that is still a pointer, it would be a very minor change that wouldn't need refactoring anywhere else.

If would seem like the point of this function is to create a copy of `nm.registeredNodes` so that outside actors aren't mutating the NodeManager's internal state, if we aren't doing a DeepCopy before returning the nodes, then outside actors can still modify the internal state.

Having looked at where this is used, the only thing we are actually using from the result of this function is the names of the objects anyway, so it is safe as it currently is, my concern would be if someone started using it to modify stuff later down the line.

It also looks like this could be simplified, we don't actually need the node object where it is being used, but that doesn't need to be addressed now.